castle/cms/cron/_social_counts.py

```diff
@@ -24,54 +22,104 @@
 import time
 import transaction
 
+USE_MULTIPROCESSING = True
+
+TOKEN_AUTH_STRING = '%%%TOKEN_AUTH%%%'
+SITE_ID_STRING = '%%%SITE_ID%%%'
+BASE_URL_STRING = '%%%BASE_URL%%%'
+
+MATOMO_TOKEN_AUTH = 'castle.matomo_token_auth'
+MATOMO_BASE_URL = 'castle.matomo_base_url'
+MATOMO_SITE_ID = 'castle.matomo_site_id'
 
 logger = logging.getLogger('castle.cms')
 
 
-def get_facebook_url_data(urls):
-    urls_string = '("%s")' % (
-        '","'.join(quote_plus(url) for url in urls)
-    )
-    access_url = COUNT_URLS['facebook']['url'] % urls_string
-    resp = requests.get(access_url).content
-    data = json.loads(resp)
-    count = 0
-    if 'data' not in data:
+def get_twitter_matomo_url_data(urls):
+    registry = getUtility(IRegistry)
+    site_id = registry.get(MATOMO_SITE_ID, None)
+    base_url = registry.get(MATOMO_BASE_URL, None)
+    token_auth = registry.get(MATOMO_TOKEN_AUTH, None)
+    if (site_id is None or site_id == '' or
+            base_url is None or base_url == '' or
+            token_auth is None or token_auth == ''):
+        return 0
+    total_count = 0
+    for url in urls:
+        query_url = COUNT_URLS['twitter_matomo']['url'].replace(
+            BASE_URL_STRING, base_url).replace(
+            SITE_ID_STRING, site_id).replace(
+            TOKEN_AUTH_STRING, token_auth) % url
+        resp = requests.get(query_url, timeout=10).content
+        datatable = json.loads(resp)
+        for d in datatable:
+            if d[u'label'] == u'twitter.com':
+                total_count += d[u'nb_visits']
+                break
+    return total_count
+
+
+def get_facebook_matomo_url_data(urls):
+    registry = getUtility(IRegistry)
+    site_id = registry.get(MATOMO_SITE_ID, None)
+    base_url = registry.get(MATOMO_BASE_URL, None)
+    token_auth = registry.get(MATOMO_TOKEN_AUTH, None)
+    if (site_id is None or site_id == '' or
+            base_url is None or base_url == '' or
+            token_auth is None or token_auth == ''):
         return 0
-    data = data['data']
-    for group in data:
-        count += group.get('like_count', 0)
-        count += group.get('total_count', 0)
-        count += group.get('share_count', 0)
-        count += group.get('comment_count', 0)
-    return count
+    total_count = 0
+    for url in urls:
+        query_url = COUNT_URLS['facebook_matomo']['url'].replace(
```

Comment:
I was trying to fit it into the mechanism you had there, using the generator methods

Reply:
It seemed like the neatest way to add Matomo support, given that it needs its own URLs structure and handling of returned values. You had that sort of thing for Facebook (though I removed it since it no longer is supported).

Reply:
yah, not sure this pattern fits very well anymore if we're doing the replacement stuff.